walnuts_cpp/nuts.hpp

```diff
@@ -300,25 +240,39 @@ void transition(Random<S, Generator>& rng,
       if (uturn_flag) break;
     }
   }
-  theta_next = span_accum.theta_select_;
+  theta_next = std::move(span_accum.theta_select_);
 }
 
-template <typename S, class F, class Generator>
+template <typename S, class F, class Generator, class H>
 void nuts(Generator& generator,
           const F& logp_grad_fun,
           const Vec<S>& inv_mass,
           S step,
           Integer max_depth,
-          const Vec<S>& theta,
-          Matrix<S>& sample) {
+          const Vec<S>& theta_init,
+          Integer num_draws,
+          H& handler) {
   Random<S, Generator> rng{generator};
-  Integer num_draws = sample.cols();
-  if (num_draws == 0) return;
-  sample.col(0) = theta;
+  Vec<S> theta = theta_init;  // copy once
+  handler(0, theta);
   for (Integer n = 1; n < num_draws; ++n) {
     transition(rng, logp_grad_fun, inv_mass, step, max_depth,
-               Vec<S>(sample.col(n - 1)), sample.col(n));
+               std::move(theta), theta);
+    handler(n, theta);
   }
 }
 
+template <typename S, class F, class Generator>
+void nuts(Generator& generator,
+          const F& logp_grad_fun,
+          const Vec<S>& inv_mass,
+          S step,
+          Integer max_depth,
+          const Vec<S>& theta_init,
+          Matrix<S>& sample) {
+  auto handler = [&sample](Integer n, const Vec<S>& v) { sample.col(n) = v; };
```

Comment:
It’s worth considering if we want to provide any other information, like the final lp value or any of the other ~7 things Stan reports

Reply:
It is, but I was setting that aside until we had the basic algorithm down.  I created an issue with a list for future reference:

https://github.com/bob-carpenter/walnuts/issues/9